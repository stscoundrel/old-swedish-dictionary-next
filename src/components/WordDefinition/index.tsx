import { capitalize } from 'lib/utils/strings'
import { lettersToRunes } from 'futhork'
import { DictionaryEntry } from 'lib/models/dictionary'
import styles from './WordDefinition.module.scss'

interface WordDefinitionProps{
  entry: DictionaryEntry,
}

export default function WordDefinition({ entry }: WordDefinitionProps) {
  const {
    headword, partOfSpeech, grammaticalAspect, information, definitions, alternativeForms,
  } = entry

  return (
    <article className={styles.section}>
      <header>
        <h1 lang="non-swe">{capitalize(headword)}</h1>

        <small className={styles.subHeading}>
          Old Swedish Dictionary - {headword.toLowerCase()}
        </small>
        <p>Meaning of Old Swedish word <em>&quot;{headword}&quot;</em> in Swedish.</p>
        <p>As defined by K.F SÃ¶derwall&apos;s dictionary of Medieval Swedish:</p>
      </header>

      {definitions.length > 1 && <p><dfn className="capitalize">{headword}</dfn> Old Swedish word can mean:</p>}
      {definitions.map((definition, index) => (
        <dl className={styles.definitionList} key={`definition-${index}`}>
          <dt><strong>{headword}</strong></dt>
          <dd
            lang="swe"
          >{definition}</dd>
        </dl>
      ))}

      {partOfSpeech.length > 0 && <p><strong>Part of speech:</strong> <em>{partOfSpeech.join(', ')}</em></p>}
      {information
        && <p><strong>Additional information:</strong> <em>{information}</em></p>
      }
      {grammaticalAspect
        && <p><strong>Grammatical aspect:</strong> <em>{grammaticalAspect}</em></p>
      }

      {alternativeForms.length > 0
        && <>
        <strong>Alternative forms or notes:</strong>
        <ul className="list">
          {alternativeForms.map((alternativeForm, index) => <li key={`alternativeForm-${index}`}>{alternativeForm}</li>)}
        </ul>
        </>
      }

      <p>Possible runic inscription in <em>Medieval Futhork</em>:
        <span className={styles.rune}>{ lettersToRunes(headword) }</span><br />
      <small>Medieval Runes were used in Sweden from 12th to 17th centuries.</small>
      </p>

    </article>
  )
}
