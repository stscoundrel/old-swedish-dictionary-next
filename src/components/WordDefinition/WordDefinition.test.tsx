import ReactDOM from 'react-dom/client'
import renderer from 'react-test-renderer'
import WordDefinition from './index'

const entry = {
  headword: 'afköra',
  partOfSpeech: 'vb',
  grammaticalAspect: 'v.',
  definitions: [
    ' , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.',
  ],
  alternativeForms: ['-bortdrifva'],
  slug: 'afkora',
}

// Optional fields artificially removed.
const minimalEntry = {
  headword: 'afköra',
  definitions: [
    ' , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.',
  ],
  partOfSpeech: '',
  grammaticalAspect: '',
  alternativeForms: [],
  slug: 'afkora',
}

describe('WordDefinition component -> full entry', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<WordDefinition entry={entry} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <WordDefinition entry={entry} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct title', () => {
    const tree = renderer.create(<WordDefinition entry={entry} />)
    const { root } = tree

    expect(root.findByType('h1').children).toEqual(['Afköra'])
  })
})

describe('WordDefinition component -> minimal entry', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(<WordDefinition entry={minimalEntry} />)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <WordDefinition entry={minimalEntry} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Has correct title', () => {
    const tree = renderer.create(<WordDefinition entry={minimalEntry} />)
    const { root } = tree

    expect(root.findByType('h1').children).toEqual(['Afköra'])
  })
})
