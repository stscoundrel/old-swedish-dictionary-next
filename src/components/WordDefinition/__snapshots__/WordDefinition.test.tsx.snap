// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WordDefinition component -> full entry Matches snapshot 1`] = `
<article
  className="section"
>
  <header>
    <h1
      lang="non-swe"
    >
      Afköra
    </h1>
    <small
      className="subHeading"
    >
      Old Swedish Dictionary - 
      afköra
    </small>
    <p>
      Meaning of Old Swedish word 
      <em>
        "
        afköra
        "
      </em>
       in Swedish.
    </p>
    <p>
      As defined by K.F Söderwall's dictionary of Medieval Swedish:
    </p>
  </header>
  <dl
    className="definitionList"
  >
    <dt>
      <strong>
        afköra
      </strong>
    </dt>
    <dd
      lang="swe"
    >
       , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.
    </dd>
  </dl>
  <p>
    <strong>
      Part of speech:
    </strong>
     
    <em>
      vb
    </em>
  </p>
  <p>
    <strong>
      Grammatical aspect:
    </strong>
     
    <em>
      v.
    </em>
  </p>
  <p>
    Possible runic inscription in 
    <em>
      Medieval Futhork
    </em>
    :
    <span
      className="rune"
    >
      ᛆᚠᚴᚯᚱᛆ
    </span>
    <br />
    <small>
      Medieval Runes were used in Sweden from 12th to 17th centuries.
    </small>
  </p>
</article>
`;

exports[`WordDefinition component -> minimal entry Matches snapshot 1`] = `
<article
  className="section"
>
  <header>
    <h1
      lang="non-swe"
    >
      Afköra
    </h1>
    <small
      className="subHeading"
    >
      Old Swedish Dictionary - 
      afköra
    </small>
    <p>
      Meaning of Old Swedish word 
      <em>
        "
        afköra
        "
      </em>
       in Swedish.
    </p>
    <p>
      As defined by K.F Söderwall's dictionary of Medieval Swedish:
    </p>
  </header>
  <dl
    className="definitionList"
  >
    <dt>
      <strong>
        afköra
      </strong>
    </dt>
    <dd
      lang="swe"
    >
       , vräka. hafwi wald han afköra  GS 43 (1416?). af ty ahren (för areno) han landboen afkörer ib.
    </dd>
  </dl>
  <p>
    Possible runic inscription in 
    <em>
      Medieval Futhork
    </em>
    :
    <span
      className="rune"
    >
      ᛆᚠᚴᚯᚱᛆ
    </span>
    <br />
    <small>
      Medieval Runes were used in Sweden from 12th to 17th centuries.
    </small>
  </p>
</article>
`;
